### PR TITLE
Updated README to include client-go<->controller diagram

### PR DESCRIPTION
**Pull request by devdattakulkarni**
The sample-controller makes extensive use of various mechanisms
available in client-go. For writing custom controllers/operators
it will be helpful if there is precise description of how the
client-go library works and how/where it interfaces with
custom controller code.

Recently we published a blog post with these details here:

https://medium.com/@cloudark/kubernetes-custom-controllers-b6c7d0668fdf

This patch includes the diagram from the post, as was recommended
by @sttts on https://github.com/kubernetes/sample-controller/issues/13

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note

```

### CODE REVIEW

File: staging/src/k8s.io/sample-controller/README.md [hunk 1]
```diff
@@ -23,6 +23,14 @@ Changes should not be made to these files manually, and when creating your own
 controller based off of this implementation you should not copy these files and
 instead run the `update-codegen` script to generate your own.
 
+## Details
+
+The sample controller uses [client-go library](https://github.com/kubernetes/client-go/tree/master/tools/cache) extensively.
+The interaction points of the controller with various mechanisms from this library are depicted in following diagram.
+
+![Alt text](./client-go-controller-interaction.jpeg.?raw=true "Kubernetes client-go and Custom Controller")
```
Comment (munnerz) [inline staging/src/k8s.io/sample-controller/README.md]: This doesn't appear to be rendering correctly for me 🤔 do relative links work for images?

Comment (munnerz) [reply staging/src/k8s.io/sample-controller/README.md]: Perhaps we can also move the image into a subdirectory too? Either `images/` or `docs/images/`?

Comment (nikhita) [reply staging/src/k8s.io/sample-controller/README.md]: Doesn't render correctly for me as well. `docs/images/` sgtm.

@devdattakulkarni Your write-up was really great as well! Would you be willing to add it here? :)

Comment (devdattakulkarni) [reply staging/src/k8s.io/sample-controller/README.md]: Thanks for the comments.

@munnerz: I will move the image to docs/images

@nikhita: Thanks for the feedback. I will include the write-up as well.

**Comment by devdattakulkarni** [comment]
@munnerz: Looks like all tests have passed..

File: staging/src/k8s.io/sample-controller/docs/controller-client-go.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+the client-go library work and their interaction points with the custom 
+controller code that you will write.
+
+<p align="center">
+  <img src="images/client-go-controller-interaction.jpeg" height="600" width="700"/>
+</p>
+
+## client-go components
+
+* Reflector: A reflector [watches the Kubernetes API](https://github.com/kubernetes/client-go/blob/master/tools/cache/reflector.go#L239) for the specified
```
Comment (sttts) [inline staging/src/k8s.io/sample-controller/docs/controller-client-go.md]: these links will unfortunately break easily. Can we mention the types with their package instead? That's more stable than line numbers.

Comment (devdattakulkarni) [reply staging/src/k8s.io/sample-controller/docs/controller-client-go.md]: Ok, makes sense. Will do that and resubmit.

**Comment by devdattakulkarni** [comment]
@sttts: I have made the suggested changes. Let me know if you have any other suggestions.

**Pull request merged**